Rebalance by default

mrtracy opened the pull request:
Three commits which enable rebalancing by default.  I think the server has reached a sufficient level of stability that this is now preferable.
- First commit adds a "balance-mode" flag, which allows users to select a non-default balancing strategy from the command line.  For instance, I will use this to select the RangeCountRebalancer when running multiple nodes on the same machine, since the usage-based balancer will not work very well in that scenario.  I'm assuming that this will also be useful in the future to test experimental balancing strategies.
- Second commit is just some mechanical renames for clarity.
- Third commit enables rebalancing by default.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/3098)

<!-- Reviewable:end -->

Commit:
Add rebalance-mode command line flag.

This command line flag allows the rebalancing mode to be modified from the
command line; instead of using the default usage-based balancer, you can select
to use a simpler range-count rebalancer. This is currently helpful for doing
rebalancing tests with multiple nodes on a single machine; in the future, it
should be useful for testing experimental rebalancing criteria.

Valid values for the flag are "usage" and "rangecount".

Commit:
Renaming for clarity in Allocator system

Some mechanical renames for clarity:

RebalancingOptions -> AllocatorOptions
"defaultBalancer" -> "usageBalancer"
"Rebalance*" -> "Balance*" in several places

Commit:
Enable Rebalancing by default

tschottdorf commented on server/context.go:
```diff
@@ -43,7 +44,8 @@ const (
 	defaultMetricsFrequency   = 10 * time.Second
 	defaultTimeUntilStoreDead = 5 * time.Minute
-	defaultAllowRebalancing   = false
+	defaultAllowRebalancing   = true
```
is this one obsolete now that there's a "none" flag?

Commit:
Remove allow-rebalancing flag completely.

Remove the "allow-rebalancing" command line flag completely; rebalancing will
now always be allowed on a running server.

Note that rebalancing is still turned off by default for most unit tests; it
must be explicitly enabled in tests which use a TestServer by setting
AllocatorOptions directly.

tschottdorf commented:
Reviewed 7 of 10 files at r2, 3 of 3 files at r4.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

<sup>**[storage/allocator.go, line 90 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3098#-K2npOd6TDCNxeFTA5_a)** ([raw file](https://github.com/cockroachdb/cockroach/blob/87c3d40005b4e822c7c9d2f42fdf907ccddf1a2d/storage/allocator.go#L90)):</sup>
this one can be a `[...]string`
<!-- Reviewable comment -K2npOd6TDCNxeFTA5_b:1315124240 -->

---

<sup>**[storage/allocator.go, line 122 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3098#-K2npW2dI-nKkwsJZkkp)** ([raw file](https://github.com/cockroachdb/cockroach/blob/87c3d40005b4e822c7c9d2f42fdf907ccddf1a2d/storage/allocator.go#L122)):</sup>
can't this go where the other pflags are defined?
<!-- Reviewable comment -K2npW2dI-nKkwsJZkkq:70825749 -->

---

<sup>**[storage/allocator.go, line 186 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3098#-K2npia_kgOD5SVFg4zG)** ([raw file](https://github.com/cockroachdb/cockroach/blob/87c3d40005b4e822c7c9d2f42fdf907ccddf1a2d/storage/allocator.go#L186)):</sup>
`default: panic()`?
<!-- Reviewable comment -K2npia_kgOD5SVFg4zH:-1108706933 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3098)
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 3 files at r4.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

<sup>**[storage/allocator.go, line 90 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3098#-K2npOd6TDCNxeFTA5_a)** ([raw file](https://github.com/cockroachdb/cockroach/blob/87c3d40005b4e822c7c9d2f42fdf907ccddf1a2d/storage/allocator.go#L90)):</sup>
+1
<!-- Reviewable comment -K2rMMH0n1cXTAlj72dw:5172523 -->

---

<sup>**[storage/allocator.go, line 122 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3098#-K2npW2dI-nKkwsJZkkp)** ([raw file](https://github.com/cockroachdb/cockroach/blob/87c3d40005b4e822c7c9d2f42fdf907ccddf1a2d/storage/allocator.go#L122)):</sup>
I think it's oK here, this is only custom implementation of `pflag.Value` we have currently.
<!-- Reviewable comment -K2rM47xX0m2VZKXPIiw:1600142624 -->

---

<sup>**[storage/allocator.go, line 186 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3098#-K2npia_kgOD5SVFg4zG)** ([raw file](https://github.com/cockroachdb/cockroach/blob/87c3d40005b4e822c7c9d2f42fdf907ccddf1a2d/storage/allocator.go#L186)):</sup>
+1, since should be impossible
<!-- Reviewable comment -K2rMCv8MuFTx7vt4D-Z:1735920426 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3098)
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: 8 of 10 files reviewed at latest revision, 3 unresolved discussions.

---

<sup>**[server/context.go, line 47 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3098#-K2nNxA7A1a3kjoF2jte-r3-47)** ([raw file](https://github.com/cockroachdb/cockroach/blob/159f2b26b595530bc5b8c373792f423fdeea6b23/server/context.go#L47)):</sup>
I have completely removed the "allow-rebalance" flag, which I think addresses this comment.
<!-- Reviewable comment -K2rmqESsrAG8SaymR_O:525874429 -->

---

<sup>**[storage/allocator.go, line 90 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3098#-K2npOd6TDCNxeFTA5_a)** ([raw file](https://github.com/cockroachdb/cockroach/blob/87c3d40005b4e822c7c9d2f42fdf907ccddf1a2d/storage/allocator.go#L90)):</sup>
Done.
<!-- Reviewable comment -K2rrvyP5c-nvvI1H3XW:-499146009 -->

---

<sup>**[storage/allocator.go, line 186 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3098#-K2npia_kgOD5SVFg4zG)** ([raw file](https://github.com/cockroachdb/cockroach/blob/87c3d40005b4e822c7c9d2f42fdf907ccddf1a2d/storage/allocator.go#L186)):</sup>
Done.
<!-- Reviewable comment -K2rs0mxQcLK3k25vHYj:-499146009 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3098)
<!-- Sent from Reviewable.io -->

tschottdorf commented:
LGTM
<!-- Reviewable comment -K2rtOPjxdfVt9LyMHod:676834233 -->

---

Reviewed 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

<sup>**[storage/allocator.go, line 90 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3098#-K2rstrruiWNVSTdy1R0)** ([raw file](https://github.com/cockroachdb/cockroach/blob/33a6f0c239cddf09f9bbb5c7c09e99c632ebd9f6/storage/allocator.go#L90)):</sup>
not a map any more. :)
<!-- Reviewable comment -K2rstrruiWNVSTdy1R1:-1831321409 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3098)
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

<sup>**[storage/allocator.go, line 90 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3098#-K2rstrruiWNVSTdy1R0)** ([raw file](https://github.com/cockroachdb/cockroach/blob/33a6f0c239cddf09f9bbb5c7c09e99c632ebd9f6/storage/allocator.go#L90)):</sup>
? Wasn't that your suggestion, to make it not a map?
<!-- Reviewable comment -K2ruMN7-alInmiMhix9:-1446896669 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3098)
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

<sup>**[storage/allocator.go, line 186 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3098#-K2npia_kgOD5SVFg4zG)** ([raw file](https://github.com/cockroachdb/cockroach/blob/87c3d40005b4e822c7c9d2f42fdf907ccddf1a2d/storage/allocator.go#L186)):</sup>
Hm. Use %s here; String() should return a strconv.Itoa'd int if the variant is missing from the map. Honestly, I'd look for a way to use Stringer to do this for you.
<!-- Reviewable comment -K2rtzcwBHhhBXAKGrnZ:1345573506 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3098)
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

<sup>**[storage/allocator.go, line 186 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3098#-K2npia_kgOD5SVFg4zG)** ([raw file](https://github.com/cockroachdb/cockroach/blob/87c3d40005b4e822c7c9d2f42fdf907ccddf1a2d/storage/allocator.go#L186)):</sup>
If it had a valid value of String(), it would be valid and would not panic here.
<!-- Reviewable comment -K2rumTzmRgBk-fydA4U:-1196921214 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3098)
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

<sup>**[storage/allocator.go, line 90 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3098#-K2rstrruiWNVSTdy1R0)** ([raw file](https://github.com/cockroachdb/cockroach/blob/33a6f0c239cddf09f9bbb5c7c09e99c632ebd9f6/storage/allocator.go#L90)):</sup>
yes, but the name is still `..Map`.
<!-- Reviewable comment -K2rvs4OwRA8M7ZGdVb2:-678930809 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3098)
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

<sup>**[storage/allocator.go, line 186 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3098#-K2npia_kgOD5SVFg4zG)** ([raw file](https://github.com/cockroachdb/cockroach/blob/87c3d40005b4e822c7c9d2f42fdf907ccddf1a2d/storage/allocator.go#L186)):</sup>
Can you re-read my comment, please?
<!-- Reviewable comment -K2rwTJ8xo2IUahQHdBh:669975248 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3098)
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: 9 of 10 files reviewed at latest revision, 2 unresolved discussions.

---

<sup>**[storage/allocator.go, line 186 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3098#-K2npia_kgOD5SVFg4zG)** ([raw file](https://github.com/cockroachdb/cockroach/blob/87c3d40005b4e822c7c9d2f42fdf907ccddf1a2d/storage/allocator.go#L186)):</sup>
Done.
<!-- Reviewable comment -K2ryqLPGtgHz57z5BDn:-499146009 -->

---

<sup>**[storage/allocator.go, line 90 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3098#-K2rstrruiWNVSTdy1R0)** ([raw file](https://github.com/cockroachdb/cockroach/blob/33a6f0c239cddf09f9bbb5c7c09e99c632ebd9f6/storage/allocator.go#L90)):</sup>
Oh, I see.  Changed to "Lookup"
<!-- Reviewable comment -K2rxpJ9sKG8x5bWN_6i:-1482472325 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3098)
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3098)
<!-- Sent from Reviewable.io -->

mrtracy commented:
Reviewed 1 of 1 files at r7.
Review status: all files reviewed at latest revision, all discussions resolved.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3098)
<!-- Sent from Reviewable.io -->

tschottdorf commented:
:shipit:

tschottdorf commented:
rebase time!
